src/Features/Core/Portable/RemoveAsyncModifier/AbstractRemoveAsyncModifierCodeFixProvider.cs

```diff
@@ -0,0 +1,236 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+using KnownTypes = Microsoft.CodeAnalysis.MakeMethodAsynchronous.AbstractMakeMethodAsynchronousCodeFixProvider.KnownTypes;
+
+namespace Microsoft.CodeAnalysis.RemoveAsyncModifier
+{
+    internal abstract class AbstractRemoveAsyncModifierCodeFixProvider<TReturnStatementSyntax, TExpressionSyntax> : SyntaxEditorBasedCodeFixProvider
+        where TReturnStatementSyntax : SyntaxNode
+        where TExpressionSyntax : SyntaxNode
+    {
+        public static readonly string EquivalenceKey = FeaturesResources.Remove_async_modifier;
+        internal sealed override CodeFixCategory CodeFixCategory => CodeFixCategory.Compile;
+
+        protected abstract bool IsAsyncSupportingFunctionSyntax(SyntaxNode node);
+        protected abstract bool TryGetExpressionBody(SyntaxNode methodSymbolOpt, out SyntaxNode expression);
+        protected abstract bool ShouldOfferFix(IMethodSymbol methodSymbol, KnownTypes knownTypes);
+        protected abstract SyntaxNode RemoveAsyncModifier(IMethodSymbol methodSymbolOpt, SyntaxNode node, KnownTypes knownTypes);
+        protected abstract SyntaxNode ConvertToBlockBody(SyntaxNode node, SyntaxNode expressionBody, SyntaxEditor editor);
+        protected abstract SyntaxNode GetLastChildOfBlock(SyntaxNode node);
+        protected abstract ControlFlowAnalysis AnalyzeControlFlow(SemanticModel semanticModel, SyntaxNode originalNode);
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var compilation = await document.Project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+            var knownTypes = new KnownTypes(compilation);
+
+            var diagnostic = context.Diagnostics.First();
+            var token = diagnostic.Location.FindToken(cancellationToken);
+            var node = token.GetAncestor(IsAsyncSupportingFunctionSyntax);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var methodSymbol = GetMethodSymbol(node, semanticModel, cancellationToken);
+
+            if (ShouldOfferFix(methodSymbol, knownTypes))
+            {
+                context.RegisterCodeFix(
+                    new MyCodeAction(c => FixAsync(document, diagnostic, c)),
+                    context.Diagnostics);
+            }
+        }
+
+        protected sealed override async Task FixAllAsync(
```

Comment:
ddo you have fix-all tests?  would like tests both with sibling methods, and with nested cases (i.e. nested lambdas, or a lambda in a method).

Reply:
All of the tests run a fix all scenario, but I'll add some with nesting and multiple methods just in case.